### PR TITLE
Proto changes for pattern template extensions

### PR DESCRIPTION
**Pull request by silverstar194**
Signed-off-by: silverstar195 <seanmaloney@google.com>

Commit Message:
Proto changes for pattern template extensions

Additional Description:
This PR is the first of a series to implement pattern matching and rewrite functionality for Envoy.

These specific proto changes:
1. Add the foundation for two extension configs
2. Change the API to expose the extension configs to end users

Additional follow up PR to implement extension can be found here and is a WIP:
[Add match and rewrite extension implementations ](https://github.com/envoyproxy/envoy/pull/22214)
[Use extensions in data and control plane](https://github.com/envoyproxy/envoy/pull/22207)

Risk Level:
Low

Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+// from the match pattern into the new path as specified by the rewrite template.
+// This is useful to allow application paths to be
+// rewritten in a way that is aware of segments with variable content like
+// identifiers. The router filter will place the original path as it was
+// before the rewrite into the :ref:`x-envoy-original-path
+// <config_http_filters_router_x-envoy-original-path>` header.
+//
+// Only one of :ref:`prefix_rewrite <envoy_v3_api_field_config.route.v3.RouteAction.prefix_rewrite>`,
+// :ref:`regex_rewrite <envoy_v3_api_field_config.route.v3.RouteAction.regex_rewrite>`,
+// or *path_template_rewrite* may be specified.
```
Comment (phlax) [inline api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:26]:
```suggestion
// or ``path_template_rewrite`` may be specified.
```

**Review by phlax — DISMISSED** [review]
@silverstar194 thanks for working on this

please ensure all literals use double backticks for correct rst rendering

File: api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+// * The path pattern ``/content/{format}/{lang}/{id}/{file}.vtt`` paired with a substitution
+//   string of ``/{lang}/{format}/{file}.vtt`` would transform ``/content/hls/en-us/12345/en_193913.vtt``
+//   into ``/en-us/hls/en_193913.vtt``.
+// [#not-implemented-hide:]
+message PatternTemplateRewriteConfig {
+  string path_template_rewrite = 1
+      [(validate.rules).string = {min_len: 1 max_len: 256 ignore_empty: true}];
+}
+
+// Holds the segments for rewriting urls base on pattern templates
```
Comment (edin44) [inline api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:61]: based

File: api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+message PatternTemplateRewrite {
+  message RewriteSegment {
+    oneof segment_type {
+      // Represents a segment of the rewritten URL, including any path segments,
+      // slash and prefix.
+      string literal = 1;
+
+      // Represents an index into the RE2 capture which value should be used
+      // to construct the rewritten URL. Note that the index should be greater
+      // than 0 as 0 index into the whole match RE2 pattern.
```
Comment (edin44) [inline api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:72]: "is the index into" or "indexes into" instead of "index into"?

**Review by edin44 — DISMISSED** [review]

File: source/common/router/config_impl.cc [hunk 4]
```diff
@@ -1382,7 +1382,8 @@ PathTemplateRouteEntryImpl::PathTemplateRouteEntryImpl(
     Server::Configuration::ServerFactoryContext& factory_context,
     ProtobufMessage::ValidationVisitor& validator)
     : RouteEntryImplBase(vhost, route, optional_http_filters, factory_context, validator),
-      path_template_(route.match().path_template()),
+      // TODO(silverstar194) Implement path template matcher
```
Comment (stevenzzzz) [inline source/common/router/config_impl.cc:1385]: the pattern is raise a unimplemented exception.

**Review by stevenzzzz — DISMISSED** [review]
LGTM. with some minor nits.

File: source/common/router/config_impl.cc (hunk 4 above)
Comment (silverstar194) [reply source/common/router/config_impl.cc:1385]: changed

File: api/envoy/config/route/v3/route_components.proto [hunk 5]
```diff
@@ -1083,7 +1056,7 @@ message RouteAction {
   // <config_http_filters_router_x-envoy-original-path>` header.
   //
   // Only one of :ref:`prefix_rewrite <envoy_v3_api_field_config.route.v3.RouteAction.prefix_rewrite>`
-  // [#comment:TODO(silverstar194) add the following once path_template_rewrite is implemented: :ref:`path_template_rewrite <envoy_v3_api_field_config.route.v3.RouteAction.path_template_rewrite>`,]
+  // :ref:`path_rewrite_policy <envoy_v3_api_field_config.route.v3.RouteAction.path_rewrite_policy>`
```
[the remark this answers is not in the document]
Comment (silverstar194) [reply api/envoy/config/route/v3/route_components.proto:1059]: changed to match others

File: api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+// Holds the segments for rewriting urls base on pattern templates
+// [#not-implemented-hide:]
+message PatternTemplateRewrite {
+  message RewriteSegment {
+    oneof segment_type {
+      // Represents a segment of the rewritten URL, including any path segments,
+      // slash and prefix.
+      string literal = 1;
+
+      // Represents an index into the RE2 capture which value should be used
```
[the remark this answers is not in the document]
Comment (silverstar194) [reply api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:70]: Editor default. Can change if needed. Couldn't find Envoy style guide. What is the recommend length?

[the remark this answers is not in the document]
Comment (phlax) [reply api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:70]: i dont think we have one for proto/rst - but i would say an ideal of 100 if not strictly enforced

File: api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+// * The pattern ``/videos/{language=lang/*}/*`` paired with a substitution string of
+//   ``/{language}`` would transform ``/videos/lang/en/video.m4s`` into ``lang/en``.
+//
+// * The path pattern ``/content/{format}/{lang}/{id}/{file}.vtt`` paired with a substitution
+//   string of ``/{lang}/{format}/{file}.vtt`` would transform ``/content/hls/en-us/12345/en_193913.vtt``
+//   into ``/en-us/hls/en_193913.vtt``.
+// [#not-implemented-hide:]
+message PatternTemplateRewriteConfig {
+  string path_template_rewrite = 1
+      [(validate.rules).string = {min_len: 1 max_len: 256 ignore_empty: true}];
```
[the remark this answers is not in the document]
Comment (silverstar194) [reply api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:58]: No reason now that it is in config proto. Will remove

[the remark this answers is not in the document]
Comment (silverstar194) [reply api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:58]: removed here as well

File: api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+//   transform ``/cat/dog`` into ``/dog/cat``.
+//
+// * The pattern ``/videos/{language=lang/*}/*`` paired with a substitution string of
+//   ``/{language}`` would transform ``/videos/lang/en/video.m4s`` into ``lang/en``.
+//
+// * The path pattern ``/content/{format}/{lang}/{id}/{file}.vtt`` paired with a substitution
+//   string of ``/{lang}/{format}/{file}.vtt`` would transform ``/content/hls/en-us/12345/en_193913.vtt``
+//   into ``/en-us/hls/en_193913.vtt``.
+// [#not-implemented-hide:]
+message PatternTemplateRewriteConfig {
```
[the remark this answers is not in the document]
Comment (silverstar194) [reply api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:56]: Would prefer to keep all the proto changes together

File: api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  string path_template_rewrite = 1
+      [(validate.rules).string = {min_len: 1 max_len: 256 ignore_empty: true}];
+}
+
+// Holds the segments for rewriting urls base on pattern templates
+// [#not-implemented-hide:]
+message PatternTemplateRewrite {
```
[the remark this answers is not in the document]
Comment (silverstar194) [reply api/envoy/extensions/pattern_template/rewrite/v3/pattern_template_rewrite.proto:62]: Will remove from this PR and move closer to extension

File: source/common/router/config_impl.cc [hunk 10]
```diff
@@ -1381,27 +1381,26 @@ PathTemplateRouteEntryImpl::PathTemplateRouteEntryImpl(
     const OptionalHttpFilters& optional_http_filters,
     Server::Configuration::ServerFactoryContext& factory_context,
     ProtobufMessage::ValidationVisitor& validator)
-    : RouteEntryImplBase(vhost, route, optional_http_filters, factory_context, validator),
-      path_template_(route.match().path_template()),
-      path_matcher_(Matchers::PathMatcher::createPattern(path_template_, !case_sensitive_)) {}
+    : RouteEntryImplBase(vhost, route, optional_http_filters, factory_context, validator) {
+  // TODO(silverstar194) Implement path template matcher
```
Comment (silverstar194) [inline source/common/router/config_impl.cc:1385]: Removed unused/to be implemented lines to meet code coverage

**Comment by htuch** [comment]
API LGTM but needs to be properly hooked up into the extension category and instance annotations and metadata.

File: api/envoy/config/route/v3/route_components.proto [hunk 11]
```diff
@@ -572,34 +572,7 @@ message RouteMatch {
...
-    //
-    // * ``/videos/*/*/*.m4s`` would match ``videos/123414/hls/1080p5000_00001.m4s``
-    //
-    // * ``/videos/{file}`` would match ``/videos/1080p5000_00001.m4s``
-    //
-    // * ``/**.mpd`` would match ``/content/123/india/dash/55/manifest.mpd``
-    // [#not-implemented-hide:]
-    string path_template = 15
-        [(validate.rules).string = {min_len: 1 max_len: 256 ignore_empty: true}];
+    core.v3.TypedExtensionConfig path_match_policy = 15;
```
[the remark this answers is not in the document]
Comment (silverstar194) [reply api/envoy/config/route/v3/route_components.proto:575]: Will add as a TODO. Can't add yet as it isn't present and will fail validation.

File: api/envoy/config/route/v3/route_components.proto [hunk 12]
```diff
@@ -1104,48 +1077,7 @@ message RouteAction {
...
-  // * The pattern ``/videos/{language=lang/*}/*`` paired with a substitution string of
-  //   ``/{language}`` would transform ``/videos/lang/en/video.m4s`` into ``lang/en``.
-  //
-  // * The path pattern ``/content/{format}/{lang}/{id}/{file}.vtt`` paired with a substitution
-  //   string of ``/{lang}/{format}/{file}.vtt`` would transform ``/content/hls/en-us/12345/en_193913.vtt``
-  //   into ``/en-us/hls/en_193913.vtt``.
-  // [#not-implemented-hide:]
-  string path_template_rewrite = 41
-      [(validate.rules).string = {min_len: 1 max_len: 256 ignore_empty: true}];
+  core.v3.TypedExtensionConfig path_rewrite_policy = 41;
```
[the remark this answers is not in the document]
Comment (silverstar194) [reply api/envoy/config/route/v3/route_components.proto:1080]: Same as above

**Review by stevenzzzz — DISMISSED** [review]
LGTM again.

**Comment by silverstar194** [comment]
@htuch Ready to merge when you approve. Just blocked on API approver

**Review by phlax — CHANGES_REQUESTED** [review]
@silverstar194 ive left a couple of docs nits inline - i didnt review the non-implemented docs as they are not currently rendered

File: api/envoy/config/route/v3/route_components.proto [hunk 13]
```diff
@@ -1045,8 +1019,8 @@ message RouteAction {
   // place the original path before rewrite into the :ref:`x-envoy-original-path
   // <config_http_filters_router_x-envoy-original-path>` header.
   //
-  // Only one of :ref:`regex_rewrite <envoy_v3_api_field_config.route.v3.RouteAction.regex_rewrite>`
-  // [#comment:TODO(silverstar194) add the following once path_template_rewrite is implemented: :ref:`path_template_rewrite <envoy_v3_api_field_config.route.v3.RouteAction.path_template_rewrite>`]
+  // Only one of :ref:`regex_rewrite <envoy_v3_api_field_config.route.v3.RouteAction.regex_rewrite>`,
+  // :ref:`path_rewrite_policy <envoy_v3_api_field_config.route.v3.RouteAction.path_rewrite_policy>`
   // or ``prefix_rewrite`` may be specified.
```
[the remark this answers is not in the document]
Comment (silverstar194) [reply api/envoy/config/route/v3/route_components.proto:1024]: What part would be a link?

[the remark this answers is not in the document]
Comment (phlax) [reply api/envoy/config/route/v3/route_components.proto:1024]: `prefix_rewrite`

[the remark this answers is not in the document]
Comment (silverstar194) [reply api/envoy/config/route/v3/route_components.proto]: changed

File: api/envoy/config/route/v3/route_components.proto [hunk 14]
```diff
@@ -1082,9 +1056,9 @@ message RouteAction {
   // before the rewrite into the :ref:`x-envoy-original-path
   // <config_http_filters_router_x-envoy-original-path>` header.
   //
-  // Only one of :ref:`prefix_rewrite <envoy_v3_api_field_config.route.v3.RouteAction.prefix_rewrite>`
-  // [#comment:TODO(silverstar194) add the following once path_template_rewrite is implemented: :ref:`path_template_rewrite <envoy_v3_api_field_config.route.v3.RouteAction.path_template_rewrite>`,]
-  // or ``regex_rewrite`` may be specified.
+  // Only one of ``regex_rewrite``, :ref:`prefix_rewrite <envoy_v3_api_field_config.route.v3.RouteAction.prefix_rewrite>`,
```
[the remark this answers is not in the document]
Comment (silverstar194) [reply api/envoy/config/route/v3/route_components.proto]: changed

File: api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto [hunk 15]
```diff
@@ -0,0 +1,44 @@
+syntax = "proto3";
+
+package envoy.extensions.pattern_template.match.v3;
```
Comment (htuch) [inline api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto:3]: Sorry, one last nit here. You've created two new extension categories, but are placing two concrete extension instances for each under a common tree. I would instead organize this as two trees:
* `envoy.extensions.path.match`
* `envoy.extensions.path.rewrite`

Then you have the two examples/default instances underneath these as:
* `envoy.extensions.path.match.pattern_template.v3.`
* `envoy.extensions.path.rewrite.pattern_template.v3`

I.e. you have two generic path match/rewrite extension categories, and the "pattern template" syntax in each case is just an instantiation of the category.

Comment (silverstar194) [reply api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto]: Sounds good. Switched over

**Comment by silverstar194** [comment]
@lizan Feel free to merge. Don't have permissions on my end

File: api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto (hunk 15 above)
Comment (htuch) [reply api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto]: Sort of - you should have pattern_template as a path component as suggested ^^ to make clear that this is one example of a "path match" or a "path rewrite".

File: api/envoy/config/route/v3/route_components.proto [hunk 16]
```diff
@@ -572,34 +572,8 @@ message RouteMatch {
...
-    //
-    // * ``/videos/*/*/*.m4s`` would match ``videos/123414/hls/1080p5000_00001.m4s``
-    //
-    // * ``/videos/{file}`` would match ``/videos/1080p5000_00001.m4s``
-    //
-    // * ``/**.mpd`` would match ``/content/123/india/dash/55/manifest.mpd``
-    // [#not-implemented-hide:]
-    string path_template = 15
-        [(validate.rules).string = {min_len: 1 max_len: 256 ignore_empty: true}];
+    // [#comment: TODO(silverstar195): Hook into extension once added]
```
Comment (lizan) [inline api/envoy/config/route/v3/route_components.proto:575]: mark these not-implemented-hide?

**Review by lizan — COMMENTED** [review]
I wasn't meant final approval, if this merges first then let's sort out the comment in proto.

**Review by phlax — DISMISSED** [review]

**Comment by phlax** [comment]
thanks for fixing doc links @silverstar194

File: api/envoy/config/route/v3/route_components.proto (hunk 16 above)
Comment (silverstar194) [reply api/envoy/config/route/v3/route_components.proto:576]: Done.

File: api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto (hunk 15 above)
Comment (silverstar194) [reply api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto]: I believe this is switched. Comment is on the old file path.

Comment (silverstar194) [reply api/envoy/extensions/pattern_template/match/v3/pattern_template_match.proto]: Ok, think I got it right now. Added the additional directory pattern_template

**Review by lizan — APPROVED** [review]

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**